src/ddscxx/src/org/eclipse/cyclonedds/pub/AnyDataWriterDelegate.cpp

```diff
@@ -124,7 +152,18 @@ AnyDataWriterDelegate::write_cdr(
         ret = dds_writecdr(writer, ser_data);
     }
 
+#ifdef DDSCXX_HAS_SHM
+    if (ret != DDS_RETCODE_OK) {
+        if (ser_data->iox_chunk != nullptr) {
+            // write cdr failed, so free the chunk
+            ret = dds_return_loan(writer, &ser_data->iox_chunk, 1);
```

Comment:
We need to double check whether return loan works with `dds_loan_shared_memory_buffer`, but it should (it is just delegated to iceoryx) There are more tests needed ...

Reply:
This should be tested indirectly with the tests in the upper layers

Agree, we might need to add some tests for these new shm APIs (at some point)